HW: Fix infinite rumble on emulation pause

Kekker-git opened the pull request:
This fixes the old issue of [rumble not stopping when emulation is paused](https://bugs.dolphin-emu.org/issues/9001). It was fixed for the GameCube adapter and LibUSB controllers [here](https://dolphin-emu.org/download/dev/59b54a77d33f2a59fabfc4d5a7643538d9b81327/), but that fix did not work for XInput or DInput controllers on Windows.

A partner and I tested on two different machines in multiple games and multiple controllers. With XInput controllers, the rumble stopped on emulation pause and resumed where appropriate, and stopped properly on emulation close. Neither me nor my partner were able to test with DInput. I have a DInput controller, but my rumble motors are not recognized by Dolphin on any build that I've tested.

I believe it follows the code style guidelines, but this is the first pull request from my team so if anything is off please let us know.

lioncash commented on Source/Core/Core/Core.cpp:
```diff
@@ -278,6 +278,10 @@ void Stop()  // - Hammertime!
   GCAdapter::ResetRumble();
 #endif
+#if defined(CIFACE_USE_XINPUT) || defined(CIFACE_USE_DINPUT)
+  for (int i = 0; i != 4; ++i)
```
This can be placed within its own static local function so it's not duplicated in another spot.

e.g.

```cpp
static void ResetRumble()
{
#if defined(__LIBUSB__)
  GCAdapter::ResetRumble();
#endif
#if defined(CIFACE_USE_XINPUT) || defined(CIFACE_USE_DINPUT)
  for (int i = 0; i < 4; ++i)
    Pad::ResetRumble(i);
#endif
}
```

lioncash commented:
This can also likely just be a single commit or two (given 2 are added, and then 2 are just reverting them). It's also a convention to prefix the commits with the general area they're modifying.

e.g.

`Core: Fix thing`

Kekker-git commented:
Yes, I figured the two reverts would seem odd. We're working on this as part of a project, and our professor wanted us to modify the "about" text with our names. I'm not entirely certain how to delete the commits or if I even can, so I just reverted them.

Sorry about the prefix bit; if anyone else is wondering, this is modifying Core/Core and Core/Core/HW.

lioncash commented:
You can amend this like so:

1. Run `git --reset hard HEAD~2` to get rid of the reversion commits
2. Now we only have 4 left to deal with so run `git rebase HEAD~4`
3. You should now be in a vim-like CLI interface. Press I (not L, in case they look identical), then remove the two lines that have first two initial commits. This will drop them entirely.
4. Change the text that says `pick` next to the  "Fix rumble on emulation pause" commit to `s` (to squash it down into the other commit).
5. Press ESC to exit editing mode and then type `:x` to save and quit.

Once this runs, type `git commit --amend". The same vim-like interface comes up, so press I and you can then edit to commit message. Exiting it is done the same way as before.

Then push the commits with `git push -f origin master`.

spycrab commented on Source/Core/DolphinWX/AboutDolphin.cpp:
```diff
@@ -36,7 +36,8 @@ AboutDolphin::AboutDolphin(wxWindow* parent, wxWindowID id, const wxString& titl
         "\n"
-        "This software should not be used to play games you do not legally own.\n");
+        "This software should not be used to play games you do not legally own.\n"
+        "\nWith contributions by: Connor Roth, Calvun Williams, Justin Futrell, Andrew Nguyen\n");
```
There is already a reference to GitHub's contributors page in the dialogue, no need to add your names to the list :P

Kekker-git commented:
It should be fixed. I squashed it to one commit, removed the reverts, and made a static ResetRumble() function that is called whenever the rumble reset is needed.

JosJuice commented on Source/Core/Core/Core.cpp:
```diff
@@ -806,6 +801,17 @@ static bool PauseAndLock(bool do_lock, bool unpause_on_unlock)
+#if defined(__LIBUSB__)
+  GCAdapter::ResetRumble();
+#endif
+#if defined(CIFACE_USE_XINPUT) || defined(CIFACE_USE_DINPUT)
```
You can use `#elif` instead of using `#endif` followed by `#if`.

leoetlino replied:
I don't think the two cases should exclude each other though, otherwise GC adapters won't be getting reset on Windows anymore.

JosJuice replied:
Oh, sorry. Yes, ignore my comment.

leoetlino commented on Source/Core/Core/Core.cpp:
```diff
@@ -806,6 +801,17 @@ static bool PauseAndLock(bool do_lock, bool unpause_on_unlock)
+static void ResetRumble()
+{
+#if defined(__LIBUSB__)
+  GCAdapter::ResetRumble();
```
The typical check is `i < 4`

leoetlino commented on Source/Core/Core/Core.cpp:
```diff
@@ -806,6 +801,17 @@ static bool PauseAndLock(bool do_lock, bool unpause_on_unlock)
   return was_unpaused;
```
You have to put this higher in the file, before it's used for the first time.

leoetlino commented on Source/Core/Core/Core.cpp:
```diff
@@ -806,6 +801,17 @@ static bool PauseAndLock(bool do_lock, bool unpause_on_unlock)
+  GCAdapter::ResetRumble();
+#endif
+#if defined(CIFACE_USE_XINPUT) || defined(CIFACE_USE_DINPUT)
+  for (int i = 0; i != 4; ++i)
```
The typical check for this is `i < 4`.

Kekker-git replied:
I'd seen in several other files that `i != 4` was used instead, which is why I used it there. Would `i < 4` be preferable?

leoetlino replied:
Using `<` for counting up to n is much more common than using `!=`, and `i != 4` is only used 3 times in the entire codebase, so yes, I think `i < 4` is preferable.

leoetlino commented on Source/Core/Core/Core.cpp:
```diff
@@ -806,6 +801,17 @@ static bool PauseAndLock(bool do_lock, bool unpause_on_unlock)
   return was_unpaused;
 }
 
+static void ResetRumble()
```
You have to put this higher in the file, before it's used for the first time.

Kekker-git replied:
I forgot to add the method to Core.h too. I'll do that.

leoetlino replied:
You don't need to (and I don't think you should) because this is a static function.

leoetlino commented:
Code looks fine to me now, but please rebase your branch and squash the commits.

Kekker-git commented:
Sorry it took me a while. I had squashed the commits, but asked my partner to move the ResetRumble() function and he didn't squash his commits. The commit has been rebased and squashed to one commit, with ResetRumble() above the Stop() function where it was first called.

JMC47 commented:
@dolphin-emu-bot rebuild
